### ISSUE TITLE
304 status catched as error

### ISSUE THREAD

**Issue by ouabbas** [flutterchina/dio#995]
### New Issue Checklist

- [ ] I have searched for a similar issue in the [project](https://github.com/flutterchina/dio/issues) and found none

### Issue Info

| Info                           | Value                                                 |      |
| ------------------------------ | ----------------------------------------------------- | ---- |
| Platform Name                  | e.g. flutter / ios / android                          |      |
| Platform Version               | e.g.  1.5.0 / 12.0 / 9.0                              |      |
| Dio Version                    | e.g.  2.1.0 / 1.0.17                                  |      |
| Android Studio / Xcode Version | e.g. Android Studio 3.3.2 / Xcode 10.2.1              |      |
| Repro rate                     | e.g. all the time (100%) / sometimes x% / only once   |      |
| Repro with our demo prj        | e.g. does it happen with our demo project?            |      |
| Demo project link              | e.g. link to a demo project that highlights the issue |      |

### Issue Description and Steps

Please fill in the detailed description of the issue (full output of any stack trace, compiler error, ...) and the steps to reproduce the issue.

**Comment by tilmanschweitzer** [comment]
I have the same issue for dio client 3.0.10. The issue is gone when I downgrade to dio 3.0.4.

Platform Name: Flutter on an iOS device
Plattform Version: Flutter 1.22.4
Dio Version: 3.0.10
Xcode Version: 12.2 (12B45b)
Repro rate: 100%

**Comment by tilmanschweitzer** [comment]
The "issue" was introduced in 3.0.10 (commit: https://github.com/flutterchina/dio/commit/55a3264c3d43d06833d95432f7866d297d7b50be
), but it seems to be a deliberate decision.

Still, such breaking changes should be mentioned in the change log. The change broke our custom caching interceptors, as they used the method `onResponse`. The control-flow gets quite awkward, when you need to use `catchError` or try-catch-blocks to deal with an expected result.

Our solution is to use a custom `validateStatus` which allows the 304 status again.

    _dio = new Dio(BaseOptions(validateStatus: (int status) {
      return status >= 200 && status < 300 || status == 304;
    }));

**Comment by nioncode** [comment]
I agree that this should have been mentioned in the changelog, but the previous implementation was broken for http servers that follow the standard, since it is not allowed to return content on a 304, which means that transforming the response in dio threw an exception.

The Readme should be updated to mention this behavior.
